smbclient/shutil.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import unicode_literals
+
+from ntpath import join, basename, splitdrive
+
+import sys
+import os.path
+from os import error as os_error
+
+from smbclient._os import remove, rmdir, listdir, stat, makedirs, readlink, symlink, scandir, copyfile_server_side
+from smbclient.path import islink, isdir
+from smbprotocol.exceptions import SMBOSError
+
+
+class Error(EnvironmentError):
+    pass
+
+
+class SpecialFileError(EnvironmentError):
+    """Raised when trying to do a kind of operation (e.g. copying) which is
+    not supported on a special file (e.g. a named pipe)"""
+
+
+def rmtree(path, ignore_errors=False, onerror=None, **kwargs):
+    """Recursively delete a directory tree.
+
+    If ignore_errors is set, errors are ignored; otherwise, if onerror
+    is set, it is called to handle the error with arguments (func,
+    path, exc_info) where func is smbclient.listdir, smbclient.remove,
+    or smbclient.rmdir;
+    path is the argument to that function that caused it to fail; and
+    exc_info is a tuple returned by sys.exc_info().  If ignore_errors
+    is false and onerror is None, an exception is raised.
+
+    :param path: The path to remove.
+    :param ignore_errors: The ignore errors flag.
+    :param onerror: The callback executed on errors
+    :param kwargs: Common arguments used to build the SMB Session.
+    :return: True if path is a dir or points to a dir.
+    """
+    if ignore_errors:
+        def onerror(*args):
+            pass
+    elif onerror is None:
+        def onerror(*args):
+            raise
+
+    try:
+        dir_entries = scandir(path, **kwargs)
+        for dir_entry in dir_entries:
+            if dir_entry.is_dir():
+                if dir_entry.is_symlink():
+                    rmdir(dir_entry.path, **kwargs)
+                else:
+                    rmtree(dir_entry.path, ignore_errors, onerror, **kwargs)
+            else:
+                try:
+                    remove(dir_entry.path, **kwargs)
+                except os_error:
+                    onerror(remove, dir_entry.path, sys.exc_info())
+    except os_error:
+        onerror(scandir, path, sys.exc_info())
+
+    try:
+        rmdir(path, **kwargs)
+    except os_error:
+        onerror(rmdir, path, sys.exc_info())
+
+
+def copy(src, dst, **kwargs):
+    """Copy data.
+
+    The destination may be a directory.
+    """
+    _check_src_dst(src, dst)
+
+    if _samefile(src, dst, **kwargs):
+        raise ValueError("`%s` and `%s` are the same file" % (src, dst))
+
+    if isdir(dst, **kwargs):
+        dst = join(dst, basename(src))
+
+    copyfile_server_side(src, dst, **kwargs)
+
+
+def copytree_copy(src, dst, symlinks=False, ignore=None, **kwargs):
```

Comment:
Should this just be `copytree`?

Reply:
Well, copytree under the hood uses copy2. I named it like that to make clear that it's not the same as shutil.copytree

Reply:
Hmm I see what you mean. I still think we should have it called copytree and just make it explicit in the docs that it just doesn’t copy the file metadata for now. We can always add that in a future release once we have a good and working copy2 implementation.

The reason why I think we should have this but not necessarily a copy2 for now is that it is better to have partial functionality whereas copy2’s only extra feature over just copy is something that I’m not 100% sure how to implement right now.

Reply:
ok, renamed to copytree and added note in the docstring